lib/ansible/modules/cloud/smartos/imgadm.py

```diff
@@ -0,0 +1,285 @@
+            changed = False
+
+        regex = 'Added "%s" image source "%s"' % (type, source)
+        if re.match(regex, stdout):
+            changed = True
+
+        # Fallthrough, assume changes
+        return (rc, stdout, errmsg(stderr), changed)
+    else:
+        # Type is ignored by imgadm(1M) here
+        cmd += ' -d %s' % (source)
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        changed = True
+
+        regex = 'Do not have image source "%s", no change' % (source)
+        if re.match(regex, stdout):
+            changed = False
+
+        regex = 'Deleted ".*" image source "%s"' % (source)
+        if re.match(regex, stdout):
+            changed = True
+
+        return (rc, stdout, errmsg(stderr), changed)
+
+def manage_images(module, present):
+    uuid = module.params['uuid']
+    pool = module.params['pool']
+    state = module.params['state']
+
+    if state == 'vacumed':
+        # Unconditionally pass '--force', otherwise we're prompted with 'y/N'
+        cmd = '{0} vacuum -f'.format(IMGADM)
+
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        if rc == 0:
+            if stdout == '':
+                changed = False
+            else:
+                changed = True
+
+        return (rc, stdout, errmsg(stderr), changed)
+
+    if present:
+        cmd = '{0} import -P {1} -q {2}'.format(IMGADM, pool, uuid)
+
+        changed = False
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        regex = 'Image {0} \(.*\) is already installed, skipping'.format(uuid)
+        if re.match(regex, stdout):
+            changed = False
+
+        regex = '.*ActiveImageNotFound.*'
+        if re.match(regex, stderr):
+            changed = False
+
+        regex = 'Imported image {0}'.format(uuid)
+        if re.match(regex, stdout):
+            changed = True
+    else:
+        cmd = '{0} delete -P {1} {2}'.format(IMGADM, pool, uuid)
+
+        changed = False
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        regex = '.*ImageNotInstalled.*'
+        if re.match(regex, stderr):
+            # Even if the 'rc' was non-zero (3), we handled the situation
+            # in order to determine if there was a change, so set rc to success.
+            rc = 0
+            changed = False
+
+        regex = 'Deleted image {0}'.format(uuid)
+        if re.match(regex, stdout):
+            changed = True
+
+    return (rc, stdout, errmsg(stderr), changed)
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            force=dict(default=None, type='bool'),
+            pool=dict(default='zones'),
+            source=dict(default=None),
+            state=dict(default=None, required=True, choices=['present', 'absent', 'deleted', 'imported', 'updated', 'vacumed']),
+            type=dict(default='imgapi', choices=['imgapi', 'docker', 'dsapi']),
+            uuid=dict(default=None)
+        ),
+        supports_check_mode=False,
```

Comment:
This module relies largely on imgadm(1M) to enforce idempotency, which does not provide a "noop" (or equivalent) mode to do a dry-run. Therefore adding check mode would require major changes to this module. It is something that could be added at some point though.